README.md
# Translation for Visual Studio Code

This extension for VSCode translates the selected string and enables the following operations.  
この VSCode 用の拡張機能は、選択した文字列を翻訳し、次の操作を可能とします。

1. Replace with translation result  
   翻訳結果での置き換え
2. Add translation results  
   翻訳結果の追加

## Usage 使用法

### １．Translate. 翻訳する

1. Select the string you want to translate.  
   翻訳したい文字列を選択します。
2. Click “Translation: Translate selection” in the context menu.(Shortcut key: Ctrl+Alt+t)  
   コンテキストメニューの「Translation: 選択範囲を翻訳」をクリックします。（ショートカットキー：Ctrl+Alt+t）
3. Move the mouse over to display the translation result.  
   マウスオバーすると、翻訳結果が表示されます。
4. Click the light bulb symbol (short bracket key: Ctrl+.), And the replace / add operation menu will be displayed. Click as necessary.  
   電球マークをクリックする（ショートカットキー：Ctrl+.）と、置換・追加の操作メニューが表示されますので、必要に応じてクリックします。

### ２．Clear translation results. 翻訳結果を消去する

1. Click “Translation: Clear translation results” in the context menu.(Shortcut key: Ctrl+Alt+x)  
   コンテキストメニューの「Translation: 翻訳結果をクリア」をクリックします。（ショートカットキー：Ctrl+Alt+x）
2. If nothing is selected, all translation results will be cleared.  
   何も選択していなければ、すべての翻訳結果がクリアされます。
3. If there is a translation result in the selected area, it will be cleared.  
   選択範囲に翻訳結果があれば、それがクリアされます。
4. If there is no translation result in the selected area, nothing happens.  
   選択範囲に翻訳結果がなければ、何も起きません。

## Requirements 必要条件

- Be able to connect to the Internet.  
  インターネットに接続できる事。

## Development source 開発ソース

<https://github.com/mako65jp/Translation.git>


CHANGELOG.md
# Change Log

## 1.0

- Display the translation result as a message.  
翻訳結果をメッセージとして表示。

## 1.1

- Added the action to add / replace the displayed translation result to the document being edited.  
表示した翻訳結果を、編集中の文書に追加／選択文字列と置き換えのアクションを追加。

## 1.2

- Added the action to copy the displayed translation result to the clipboard.  
表示した翻訳結果を、クリップボードにコピーのアクションを追加。

- Change the displayed translation result action.  
表示した翻訳結果のアクションを変更。
- Add to the end of the line with the selected text of the document being edited.  
編集中の文書の選択文字列のある行末に追加。
- Add a comment line to the line below the selected string.  
選択文字列の下の行にコメント行として追加。

## 1.3

- Fixed a bug when not connected to the Internet.  
インターネット非接続時の不具合を修正。

## 1.4

- Added extended function settings.  
拡張機能設定を追加。

## 1.5

- Refactoring the translation process.  
翻訳処理のリファクタリング。
- The translation result has been changed to Diagnostics display.  
翻訳結果は、Diagnostics 表示に変更。
- Actions on translation results are abolished.  
翻訳結果に対するアクションは廃止。

## 2.0

- Large scale refactoring.  
大規模リファクタリング。
- Implement action (replacement only) on translation result with code action.  
翻訳結果に対するアクション（置換のみ）をコードアクションで実装。
- Implemented the translation result clear function.  
翻訳結果クリア機能を実装。

## 2.1

- Implemented "Add Below" action for translation results.  
翻訳結果に対するアクションに、「下に追加」を実装。

### 2.1.2

- Improved language recognition errors in selected areas.  
選択範囲の言語認識ミスを改善。

## 2.2

- Added a function to display a warning when the file is opened when it differs from the default encoding.  
ファイルオープン時に、デフォルトエンコードと異なる場合には、警告を表示する機能を付加。


test.py
Extension Anatomy
In the last topic, you were able to get a basic extension running. How does it work under the hood?

The Hello World extension does 3 things:

Registers the onCommand Activation Event: onCommand:extension.helloWorld, so the extension becomes
activated when user runs the Hello World command.
Uses the contributes.commands Contribution Point to make the command Hello World available in the
Command Palette, and bind it to a command ID extension.helloWorld.
Uses the commands.registerCommand VS Code API to bind a function to the registered command ID
extension.helloWorld.
テキストを整形する。
翻訳結果を表示し処置を促す。
Display translation results and prompt treatment.
翻訳結果を表示し、処理を促します。
Displays the translation result and prompts processing.

翻訳結果を追加する
Add translation result
// translation :: 翻訳結果を追加

翻訳結果で置き換える。
Replace with translation results.

何もせず閉じる。
Close without doing anything.

翻訳コマンドの実装

クリップボードにコピーする。


src/controller.ts
'use strict';

import * as vscode from 'vscode';
import * as events from 'events';
import Translator from './translator';
import { basename } from 'path';
import { readFile } from 'fs';

export default class Controller implements vscode.Disposable {
    private _translator: Translator = Object.create(null);
    private _enabled: boolean = false;

    private _extensionContext: vscode.ExtensionContext;
    private _eventEmitter: events.EventEmitter = new events.EventEmitter();

    static TranslationCommand: string = 'translation.translation';
    static ClearCommand: string = 'translation.clear';

    private registerCommand(command: string) {
        const self = this;

        this._extensionContext.subscriptions.push(
            vscode.commands.registerCommand(command, () => {
                self._eventEmitter.emit(command);
            })
        );
    }

    private registerTextEditorCommand(command: string) {
        const self = this;

        this._extensionContext.subscriptions.push(
            vscode.commands.registerTextEditorCommand(command, editor => {
                self._eventEmitter.emit(command, editor);
            })
        );
    }

    // Selection translation command
    private translation(editor: vscode.TextEditor) {
        this.enableTranslator();
        this._translator.translation(editor);
    }

    // Translation result clear command
    private clearTranslations(editor: vscode.TextEditor) {
        this.enableTranslator();
        this._translator.clearTranslations(editor);
    }

    // Enable translation processing
    private enableTranslator() {
        if (!this._enabled) {
            this._translator = new Translator();
            this._translator.activate();
            this._enabled = true;
        }
    }

    constructor(context: vscode.ExtensionContext) {
        this._extensionContext = context;
    }

    activate() {
        const self = this;

        // Registering commands.
        this.registerTextEditorCommand(Controller.TranslationCommand);
        this.registerTextEditorCommand(Controller.ClearCommand);

        // Registering event handlers.
        this._eventEmitter.on(Controller.TranslationCommand, editor => {
            if (editor && !editor.selection.isEmpty) {
                self.translation(editor);
            }
        });
        this._eventEmitter.on(Controller.ClearCommand, editor => {
            if (editor) {
                self.clearTranslations(editor);
            }
        });

        vscode.workspace.onDidOpenTextDocument(textDocument => {
            const editorText = textDocument.getText();
            const editorEncoding = vscode.workspace.getConfiguration('files').encoding;

            readFile(textDocument.fileName, (err, data) => {
                if (err) {
                    return;
                }

                console.log('');
                console.log(textDocument.fileName);

                // 表示しているテキストをデコードして、エンコードを推測してみる方法。
                self.guessEncoding(data, editorText, editorEncoding, textDocument.fileName);
            });
        });

        vscode.window.onDidChangeActiveTextEditor(() => {
            vscode.commands.executeCommand('notifications.hideToasts');
        });
    }

    dispose() {
        if (this._translator) {
            this._translator.dispose();
        }
    }

    // 表示しているテキストをデコードして、エンコードを推測してみる方法。
    guessEncoding(array: Uint8Array, editorText: string, editorEncoding: string, fileName: string) {
        const encoding = require('encoding-japanese');
        editorEncoding = editorEncoding
            .toUpperCase()
            .replace('-', '')
            .replace('_', '')
            .replace('SHIFT', 'S');

        //encoding --> UTF8
        //encoding --> EUCJP
        //encoding --> SJIS
        //encoding --> ASCII
        const result = encoding
            .detect(array)
            .toUpperCase()
            .replace('-', '')
            .replace('_', '')
            .replace('SHIFT', 'S');
        console.log('encoding --> ' + result);

        // Since it is a single byte file, no notification is required.
        if (result === 'ASCII') {
            return;
        }

        // If the encoding is UTF-8, check if BOM is included.
        let isBOM = false;
        if (result === 'UTF8' && array.length >= 3) {
            if (array[0] === 0xef && array[1] === 0xbb && array[2] === 0xbf) {
                isBOM = true;
            }
        }

        // No notification required as it matches the editor's recognition.
        if (isBOM === false && result === editorEncoding) {
            return;
        }

        const resultEncoding = isBOM ? result + ' with BOM' : result;
        const message = 'The file "' + basename(fileName) + '" is encoded in "' + resultEncoding + '". ';
        vscode.window.showWarningMessage(message);
        return;
    }
}


src/translation.ts
import * as vscode from 'vscode';
var request = require('request-promise');

// Get the character string selected on the active editor
export function getSelectedText(
  document: vscode.TextDocument,
  selection: vscode.Selection
): string {
  if (!document || selection.isEmpty) {
    return '';
  }

  const selectedText = document.getText(selection);

  // Erase consecutive whitespace.
  // Remove leading and trailing blanks.
  return selectedText.replace(/(\s)+/g, ' ').replace(/(^\s+)|(\s+$)/g, '');
}

export async function translation(text: string) {
  // Get translation pattern settings.
  const pattern = vscode.workspace
    .getConfiguration('translationExt')
    .get<any[]>('pattern') || [
    { src: 'en', target: ['ja'] },
    { src: 'ja', target: ['en'] }
  ];
  const target = pattern[0].target[0] || '';

  // Translate using the first pattern.
  return await requestWithTimeout2(text, target).then(async results => {
    if (results.status === 0) {
      // There was a response from the translation site.

      if (results.res.src === target) {
        // If the source language and the specified language are equal.

        // Translate again with the next translation pattern setting.
        const next = pattern.find(v => v.src === results.res.src);
        if (next && next.target[0]) {
          // The following translation pattern was found.
          return await requestWithTimeout2(text, next.target[0]).then(
            results => {
              if (results.status === 0) {
                // Translated correctly.
                return parsTranslated(
                  results.res,
                  results.res.src,
                  next.target[0]
                );
              } else {
                // There was no response from the translation site.
                return parsTranslated(null);
              }
            }
          );
        } else {
          // The next conversion pattern was not found.
          return parsTranslated(null);
        }
      }
      // Translated correctly.
      return parsTranslated(results.res, results.res.src, target);
    } else {
      // Fatal error.
      return parsTranslated(null);
    }
  });
}

// Format the translation results.
function parsTranslated(res: any, origLang?: string, transLang?: string) {
  // When translation results cannot be obtained.
  if (!res || !res.sentences) {
    return {
      trans: '',
      orig: '',
      lang: { trans: '', orig: '' }
    };
  }

  // When translation results are obtained.
  return {
    trans: convertTranslatedText(res.sentences),
    orig: convertOriginalText(res.sentences),
    lang: { trans: transLang + '', orig: origLang + '' }
  };
}

// Convert the translation result into a string.
function convertTranslatedText(sentences: []) {
  let result: string[] = [];
  sentences.forEach((s: { orig: string; trans: string }) => {
    result.push(s.trans || '');
  });
  return result.join('\n');
}

// Convert the translation source to a string.
function convertOriginalText(sentences: []) {
  let result: string[] = [];
  sentences.forEach((s: { orig: string; trans: string }) => {
    result.push(s.orig || '');
  });
  return result.join('\n');
}

// Generate translation site parameters.
function getParameterForTranslateSite(text: string, target: string) {
  const configProxy = String(
    vscode.workspace.getConfiguration().get('http.proxy')
  );

  let options = {
    uri:
      'https://translate.google.com/translate_a/single?client=gtx' +
      '&dt=t&dt=bd&ie=UTF-8&oe=UTF-8&dj=1&source=icon' +
      '&sl=auto' +
      '&tl=' +
      (target || 'auto') +
      '&q=' +
      encodeURIComponent(text),
    proxy: String(vscode.workspace.getConfiguration().get('http.proxy')),
    json: true
  };
  return options;
}

async function requestWithTimeout2(text: string, target: string) {
  // Generate translation site parameters.
  const options = getParameterForTranslateSite(text, target);

  // Get translation processing timeout setting.
  const processingTimeout = vscode.workspace
    .getConfiguration('translationExt')
    .get<number>('processingTimeout', 750);

  return await Promise.race([
    request(options),
    timeout(processingTimeout)
  ]).then(
    async res => {
      if (res.src) {
        return { status: 0, res: res };
      } else {
        return { status: 900, res: '' };
      }
    },
    reason => {
      return { status: 900, res: reason };
    }
  );
}

async function timeout(msec: number) {
  return new Promise((_, reject) => setTimeout(reject, msec));
}


src/diagnosticsAndActions.ts
'use strict';

import * as vscode from 'vscode';
import * as utils from './utils';
import Translator from './translator';

interface Map<V> {
    [key: string]: V;
}

export default class DiagnosticsAndActions implements vscode.CodeActionProvider, vscode.Disposable {
    private _diagnosticMap: Map<vscode.Diagnostic[]> = {};
    private _diagnostics: vscode.DiagnosticCollection = Object.create(null);

    provideCodeActions(
        document: vscode.TextDocument,
        range: vscode.Range,
        context: vscode.CodeActionContext,
        token: vscode.CancellationToken
    ): vscode.Command[] {
        const diagnostic = context.diagnostics.filter((d, i) => d.source === Translator.Source)[0];
        const suggestion = diagnostic.message;
        const error = document.getText(range);

        let commands: vscode.Command[] = [];
        commands.push(
            {
                title: "Replace with '" + suggestion.substring(0, 80) + " ... '",
                command: Translator.ReplaceOnSuggestion,
                arguments: [document, diagnostic, error, suggestion]
            },
            {
                title: "Add with '" + suggestion.substring(0, 80) + " ... '",
                command: Translator.AddOnSuggestion,
                arguments: [document, diagnostic, error, suggestion]
            }
        );

        return commands;
    }

    dispose() {}

    private fixWithSuggestion(document: vscode.TextDocument, diagnostic: vscode.Diagnostic, error: string, suggestion: string): any {
        let diagnostics: vscode.Diagnostic[] = this._diagnosticMap[document.uri.toString()];
        let index: number = diagnostics.indexOf(diagnostic);

        diagnostics.splice(index, 1);

        this._diagnosticMap[document.uri.toString()] = diagnostics;
        this._diagnostics.set(document.uri, diagnostics);

        let edit = new vscode.WorkspaceEdit();
        edit.replace(document.uri, diagnostic.range, suggestion);
        return vscode.workspace.applyEdit(edit);
    }

    public createDiagnostics(editor: vscode.TextEditor, souece: string, transText: string) {
        const document = editor.document;
        const range: vscode.Range = editor.selection;

        let diagnostics: vscode.Diagnostic[] = (this._diagnostics.get(document.uri) || []).filter(d => !range.intersection(d.range));

        diagnostics.push(this.createDiagnosticObject(document, range, souece, vscode.DiagnosticSeverity.Information, transText));

        this._diagnostics.set(document.uri, diagnostics);
        this._diagnosticMap[document.uri.toString()] = diagnostics;
    }

    private createDiagnosticObject(
        document: vscode.TextDocument,
        range: vscode.Range,
        message: string,
        severity: vscode.DiagnosticSeverity,
        suggest: string
    ) {
        message = message.replace(/\r?\n/g, '');
        suggest = suggest.replace(/\r?\n/g, '');
        if (!utils.isEmptyOrSpaces(suggest)) {
            message = `${message}\n${suggest}`;
        }

        // let startPos = document.positionAt(pos[0]);
        // let endPos = document.positionAt(pos[1]);
        // let range = new vscode.Range(startPos, endPos);
        let diag = new vscode.Diagnostic(range, message, severity);
        return diag;
    }
}


src/translator.ts
'use strict';

import * as vscode from 'vscode';
// import * as DiagnosticsAndActions from './diagnosticsAndActions';
import * as utils from './utils';
var request = require('request-promise');

interface Map<V> {
    [key: string]: V;
}

export default class Translator implements vscode.CodeActionProvider, vscode.Disposable {
    private _disposable: vscode.Disposable = Object.create(null);

    private _isEnabled = false;

    private static ConfigProxy = vscode.workspace.getConfiguration().get<string>('http.proxy') || '';

    public static Source = 'Translator';
    private static Config: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration(Translator.Source);

    private _diagnosticMap: Map<vscode.Diagnostic[]> = {};
    private _diagnostics: vscode.DiagnosticCollection = Object.create(null);

    public static ReplaceOnSuggestion: string = Translator.Source + '.replaceOnSuggestion';
    private _replaceOnSuggestionCommand: vscode.Disposable = Object.create(null);

    public static AddOnSuggestion: string = Translator.Source + '.addOnSuggestion';
    private _addOnSuggestionCommand: vscode.Disposable = Object.create(null);

    public provideCodeActions(
        document: vscode.TextDocument,
        range: vscode.Range,
        context: vscode.CodeActionContext,
        token: vscode.CancellationToken
    ): vscode.Command[] {
        const diagnostic = context.diagnostics.filter((d, i) => d.source === Translator.Source)[0];
        const suggestion = diagnostic.message;
        const error = document.getText(range);

        let commands: vscode.Command[] = [];
        commands.push(
            {
                title: "置換： '" + suggestion.substring(0, 80) + " ... '",
                command: Translator.ReplaceOnSuggestion,
                arguments: [document, diagnostic, error, suggestion]
            },
            {
                title: "下に追加： '" + suggestion.substring(0, 80) + " ... '",
                command: Translator.AddOnSuggestion,
                arguments: [document, diagnostic, error, suggestion]
            }
        );

        return commands;
    }

    private replaceWithSuggestion(document: vscode.TextDocument, diagnostic: vscode.Diagnostic, error: string, suggestion: string): any {
        let diagnostics: vscode.Diagnostic[] = this._diagnosticMap[document.uri.toString()];
        let index: number = diagnostics.indexOf(diagnostic);

        diagnostics.splice(index, 1);

        this._diagnosticMap[document.uri.toString()] = diagnostics;
        this._diagnostics.set(document.uri, diagnostics);

        let edit = new vscode.WorkspaceEdit();
        edit.replace(document.uri, diagnostic.range, suggestion);
        vscode.workspace.applyEdit(edit);
    }

    private addWithSuggestion(document: vscode.TextDocument, diagnostic: vscode.Diagnostic, error: string, suggestion: string): any {
        const editor = vscode.window.activeTextEditor;
        if (editor) {
            let diagnostics: vscode.Diagnostic[] = this._diagnosticMap[document.uri.toString()];
            let index: number = diagnostics.indexOf(diagnostic);

            const anchor = new vscode.Position(diagnostics[index].range.end.line + 1, 0);
            const active = new vscode.Position(diagnostics[index].range.end.line + 1, 0);
            const selection = new vscode.Selection(anchor, active);
            editor.selection = selection;

            editor
                .edit(editBuilder => {
                    editBuilder.replace(editor.selection, '[Translation]\n' + suggestion + '\n');
                })
                .then(success => {
                    if (success) {
                        diagnostics.splice(index, 1);

                        this._diagnosticMap[document.uri.toString()] = diagnostics;
                        this._diagnostics.set(document.uri, diagnostics);
                        vscode.commands.executeCommand('editor.action.addCommentLine');
                    }
                });
        }
    }

    private static createDiagnosticObject(range: vscode.Range, suggest: string) {
        let diag = new vscode.Diagnostic(range, suggest, vscode.DiagnosticSeverity.Information);
        diag.source = Translator.Source;
        return diag;
    }

    private createDiagnostics(editor: vscode.TextEditor, transText: string) {
        const document = editor.document;
        const range: vscode.Range = editor.selection;

        let diagnostics: vscode.Diagnostic[] = (this._diagnostics.get(document.uri) || []).filter(d => !range.intersection(d.range));

        diagnostics.push(Translator.createDiagnosticObject(range, transText));

        this._diagnostics.set(document.uri, diagnostics);
        this._diagnosticMap[document.uri.toString()] = diagnostics;
    }

    public async translation(editor: vscode.TextEditor) {
        if (!this._isEnabled) {
            return;
        }

        const selectionText = utils.getSelectedText(editor.document, editor.selection);
        if (!selectionText) {
            return;
        }

        await this.postTranslation(selectionText).then(results => {
            if (results.trans) {
                // If you can translate
                if (results.orig !== results.trans) {
                    this.createDiagnostics(editor, results.trans);
                }
            } else {
                // If translation results cannot be obtained or if an error occurs,
                // a message is displayed in the lower right.
                vscode.window.showErrorMessage('Translation ----->>> ERROR');
            }
        });
    }

    private async postTranslation(text: string) {
        // Get translation pattern settings.
        const pattern = Translator.Config.get<any[]>('pattern') || [{ src: 'en', target: ['ja'] }, { src: 'ja', target: ['en'] }];
        const target = pattern[0].target[0] || '';

        // Translate using the first pattern.
        return await this.requestWithTimeout(text, target).then(async results => {
            if (results && results.status === 0) {
                // There was a response from the translation site.
                let isSource = results.res.src === target;
                if (!isSource) {
                    results.res.ld_result.srclangs.forEach((srclang: string) => {
                        if (srclang === target) {
                            isSource = true;
                        }
                    });
                }
                if (isSource) {
                    // If the source language and the specified language are equal.

                    // Translate again with the next translation pattern setting.
                    const next = pattern.find(v => v.src === target);
                    if (next && next.target[0]) {
                        // The following translation pattern was found.
                        return await this.requestWithTimeout(text, next.target[0]).then(results => {
                            if (results.status === 0) {
                                // Translated correctly.
                                return this.parsTranslated(results.res, results.res.src, next.target[0]);
                            } else {
                                // There was no response from the translation site.
                                return this.parsTranslated(null);
                            }
                        });
                    } else {
                        // The next conversion pattern was not found.
                        return this.parsTranslated(null);
                    }
                }
                // Translated correctly.
                return this.parsTranslated(results.res, results.res.src, target);
            } else {
                // Fatal error.
                return this.parsTranslated(null);
            }
        });
    }

    private async requestWithTimeout(text: string, target: string) {
        // Generate translation site parameters.
        const options = this.getParameterForTranslateSite(text, target);

        // Get translation processing timeout setting.
        const processingTimeout = Translator.Config.get<number>('processingTimeout', 750);

        return await Promise.race([request(options), new Promise((_, reject) => setTimeout(reject, processingTimeout))]).then(
            async res => {
                if (res.src) {
                    return { status: 0, res: res };
                } else {
                    return { status: 999, res: '' };
                }
            },
            reason => {
                return { status: 999, res: reason };
            }
        );
    }

    // Generate translation site parameters.
    private getParameterForTranslateSite(text: string, target: string) {
        let options = {
            uri:
                'https://translate.google.com/translate_a/single?client=gtx' +
                '&dt=t&dt=bd&ie=UTF-8&oe=UTF-8&dj=1&source=icon' +
                '&sl=auto' +
                '&tl=' +
                (target || 'auto') +
                '&q=' +
                encodeURIComponent(text),
            proxy: Translator.ConfigProxy,
            json: true
        };
        return options;
    }

    // Format the translation results.
    private parsTranslated(res: any, origLang?: string, transLang?: string) {
        // When translation results cannot be obtained.
        if (!res || !res.sentences) {
            return {
                trans: '',
                orig: '',
                lang: { trans: '', orig: '' }
            };
        }

        // When translation results are obtained.
        return {
            trans: this.convertTranslatedText(res.sentences),
            orig: this.convertOriginalText(res.sentences),
            lang: { trans: transLang + '', orig: origLang + '' }
        };
    }

    // Convert the translation result into a string.
    private convertTranslatedText(sentences: []) {
        let result: string[] = [];
        sentences.forEach((s: { orig: string; trans: string }) => {
            result.push(s.trans || '');
        });
        return result.join('\n');
    }

    // Convert the translation source to a string.
    private convertOriginalText(sentences: []) {
        let result: string[] = [];
        sentences.forEach((s: { orig: string; trans: string }) => {
            result.push(s.orig || '');
        });
        return result.join('\n');
    }

    public clearTranslations(editor: vscode.TextEditor) {
        const document = editor.document;
        let diagnostics: vscode.Diagnostic[];

        if (editor.selection.isEmpty) {
            diagnostics = [];
        } else {
            const range: vscode.Range = editor.selection;
            diagnostics = (this._diagnostics.get(document.uri) || []).filter(d => !range.intersection(d.range));
        }

        this._diagnostics.set(document.uri, diagnostics);
        this._diagnosticMap[document.uri.toString()] = diagnostics;
    }

    activate() {
        const self = this;

        this._replaceOnSuggestionCommand = vscode.commands.registerCommand(
            Translator.ReplaceOnSuggestion,
            this.replaceWithSuggestion.bind(this)
        );
        this._addOnSuggestionCommand = vscode.commands.registerCommand(Translator.AddOnSuggestion, this.addWithSuggestion.bind(this));

        let subscriptions: vscode.Disposable[] = [];

        // vscode.window.onDidChangeActiveTextEditor(
        //   this.onActiveTextEditorChanged,
        //   this,
        //   subscriptions
        // );

        this._diagnostics = vscode.languages.createDiagnosticCollection(Translator.Source);

        vscode.workspace.onDidCloseTextDocument(
            textDocument => {
                self._diagnostics.delete(textDocument.uri);
            },
            null,
            subscriptions
        );

        vscode.languages.registerCodeActionsProvider('*', this);

        this._disposable = vscode.Disposable.from(...subscriptions);

        this._isEnabled = true;
    }

    dispose() {
        this._replaceOnSuggestionCommand.dispose();
        this._addOnSuggestionCommand.dispose();
        this._disposable.dispose();
    }
}
